app/tests/ble/central/src/main.c

```diff
@@ -0,0 +1,370 @@
+                           const void *data, uint16_t length) {
+    if (!data) {
+        LOG_DBG("[UNSUBSCRIBED]");
+        params->value_handle = 0U;
+        return BT_GATT_ITER_STOP;
+    }
+
+    LOG_HEXDUMP_DBG(data, length, "payload");
+
+    return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t discover_func(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+                             struct bt_gatt_discover_params *params) {
+    int err;
+
+    if (!attr) {
+        LOG_DBG("Discover complete");
+        (void)memset(params, 0, sizeof(*params));
+        return BT_GATT_ITER_STOP;
+    }
+
+    LOG_DBG("[ATTRIBUTE] handle %u", attr->handle);
+
+    if (!bt_uuid_cmp(discover_params.uuid, BT_UUID_HIDS)) {
+        memcpy(&uuid, BT_UUID_HIDS_REPORT, sizeof(uuid));
+        discover_params.uuid = &uuid.uuid;
+        discover_params.start_handle = attr->handle + 1;
+        discover_params.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+
+        err = bt_gatt_discover(conn, &discover_params);
+        if (err) {
+            LOG_DBG("Discover failed (err %d)", err);
+        }
+    } else if (!bt_uuid_cmp(discover_params.uuid, BT_UUID_HIDS_REPORT)) {
+        memcpy(&uuid, BT_UUID_GATT_CCC, sizeof(uuid));
+        discover_params.uuid = &uuid.uuid;
+        discover_params.start_handle = attr->handle + 2;
+        discover_params.type = BT_GATT_DISCOVER_DESCRIPTOR;
+        subscribe_params.value_handle = bt_gatt_attr_value_handle(attr);
+
+        err = bt_gatt_discover(conn, &discover_params);
+        if (err) {
+            LOG_DBG("Discover failed (err %d)", err);
+        }
+    } else {
+        subscribe_params.notify = notify_func;
+        subscribe_params.value = BT_GATT_CCC_NOTIFY;
+        subscribe_params.ccc_handle = attr->handle;
+
+        err = bt_gatt_subscribe(conn, &subscribe_params);
+        if (err && err != -EALREADY) {
+            LOG_DBG("Subscribe failed (err %d)", err);
+        } else {
+            LOG_DBG("[SUBSCRIBED]");
+        }
+
+        return BT_GATT_ITER_STOP;
+    }
+
+    return BT_GATT_ITER_STOP;
+}
+
+static void reconnect(const bt_addr_le_t *addr) {
+    struct bt_le_conn_param *param;
+    int err = bt_le_scan_stop();
+    if (err < 0) {
+        LOG_DBG("Stop LE scan failed (err %d)", err);
+    }
+
+    param = BT_LE_CONN_PARAM_DEFAULT;
+    err = bt_conn_le_create(addr, BT_CONN_LE_CREATE_CONN, param, &default_conn);
+    if (err < 0) {
+        LOG_DBG("Create conn failed (err %d)", err);
+        start_scan();
+    }
+}
+
+static bool eir_found(struct bt_data *data, void *user_data) {
+    bt_addr_le_t *addr = user_data;
+    int i;
+
+    LOG_DBG("[AD]: %u data_len %u", data->type, data->data_len);
+
+    switch (data->type) {
+    case BT_DATA_UUID16_SOME:
+    case BT_DATA_UUID16_ALL:
+        if (data->data_len % sizeof(uint16_t) != 0U) {
+            LOG_DBG("AD malformed");
+            return true;
+        }
+
+        for (i = 0; i < data->data_len; i += sizeof(uint16_t)) {
```

Comment:
This can read past the end of the data unless we verify that data_len is a multiple of sizeof(uint16_t). Consider `data_len == 1`. On the first iteration, `0 < 1` so we enter the loop, and then we memcpy 2 bytes out of an array with only 1 byte.

Reply:
To be honest... Given we're only using this for tests, I'm fine with the simplicity this affords us and if we crash later, I'll bang my head over it then.

Reply:
I see now that there is a check for the data length. If that was always there and you didn't just add that, then I must have just skimmed over it.